tools/generate-helpers/pg-table-bindings/store.go.tpl

```diff
@@ -357,14 +359,22 @@ func (s *storeImpl) Upsert(ctx context.Context, obj *{{.Type}}) error {
     } else if !ok {
         return sac.ErrResourceAccessDenied
     }
-    {{- else if eq .ResourceType "globallyScoped" }}
+    {{- else if .Obj.IsGloballyScoped }}
     scopeChecker := sac.GlobalAccessScopeChecker(ctx).AccessMode(storage.Access_READ_WRITE_ACCESS).Resource(targetResource)
+    {{- else if and (.Obj.IsDirectlyScoped) (.Obj.IsClusterScope) }}
+    scopeChecker := sac.GlobalAccessScopeChecker(ctx).AccessMode(storage.Access_READ_WRITE_ACCESS).Resource(targetResource).
+        ClusterID({{ "obj" | .Obj.GetClusterID }})
+    {{- else if and (.Obj.IsDirectlyScoped) (.Obj.IsNamespaceScope) }}
+    scopeChecker := sac.GlobalAccessScopeChecker(ctx).AccessMode(storage.Access_READ_WRITE_ACCESS).Resource(targetResource).
+        ClusterID({{ "obj" | .Obj.GetClusterID }}).Namespace({{ "obj" | .Obj.GetNamespace }})
+    {{- end }}
```

Comment:
It might be useful later on to move the scopeChecker extraction to a dedicated function rather than duplicating the drill-down to the right ScopeCheckerCore.
An alternative approach to scope checks could be (as used in UpsertMany) to extract the effective access scope and check that the input (or output for read) object is in scope.

Reply:
Extracted to a template

Reply:
https://github.com/stackrox/stackrox/pull/1320